File: binding/src/libqjsonobject.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qjsonobject.html
pub const qjsonobject = struct {

    /// New constructs a new QJsonObject object.
    pub fn New() QtC.QJsonObject {
        return qtc.QJsonObject_new();
    }


    /// New2 constructs a new QJsonObject object.
    pub fn New2(other: ?*anyopaque) QtC.QJsonObject {
        return qtc.QJsonObject_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonObject_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonObject_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn FromVariantMap(mapVal: *const QMap<QString, QVariant>) QtC.QJsonObject {
        return qtc.QJsonObject_FromVariantMap(@ptrCast(mapVal));
    }

    pub fn ToVariantMap(self: ?*anyopaque, ) QMap<QString, QVariant> {
        return qtc.QJsonObject_ToVariantMap(@ptrCast(self));
    }

    pub fn FromVariantHash(mapVal: *const QHash<QString, QVariant>) QtC.QJsonObject {
        return qtc.QJsonObject_FromVariantHash(@ptrCast(mapVal));
    }

    pub fn ToVariantHash(self: ?*anyopaque, ) QHash<QString, QVariant> {
        return qtc.QJsonObject_ToVariantHash(@ptrCast(self));
    }

    pub fn Keys(self: ?*anyopaque, ) []const u8 {
        return qtc.QJsonObject_Keys(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QJsonObject_Size(@ptrCast(self));
    }

    pub fn Count(self: ?*anyopaque, ) i64 {
        return qtc.QJsonObject_Count(@ptrCast(self));
    }

    pub fn Length(self: ?*anyopaque, ) i64 {
        return qtc.QJsonObject_Length(@ptrCast(self));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QJsonObject_IsEmpty(@ptrCast(self));
    }

    pub fn Value(self: ?*anyopaque, key: []const u8) QtC.QJsonValue {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_Value(@ptrCast(self), key_str);
    }

    pub fn OperatorSubscript(self: ?*anyopaque, key: []const u8) QtC.QJsonValue {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_OperatorSubscript(@ptrCast(self), key_str);
    }

    pub fn Remove(self: ?*anyopaque, key: []const u8) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QJsonObject_Remove(@ptrCast(self), key_str);
    }

    pub fn Take(self: ?*anyopaque, key: []const u8) QtC.QJsonValue {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_Take(@ptrCast(self), key_str);
    }

    pub fn Contains(self: ?*anyopaque, key: []const u8) bool {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_Contains(@ptrCast(self), key_str);
    }

    pub fn Begin(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonObject_Begin(@ptrCast(self)));
    }

    pub fn Begin2(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject_Begin2(@ptrCast(self));
    }

    pub fn ConstBegin(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject_ConstBegin(@ptrCast(self));
    }

    pub fn End(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonObject_End(@ptrCast(self)));
    }

    pub fn End2(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject_End2(@ptrCast(self));
    }

    pub fn ConstEnd(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject_ConstEnd(@ptrCast(self));
    }

    pub fn Erase(self: ?*anyopaque, it: *It::value_type) It::value_type {
        return @ptrCast(qtc.QJsonObject_Erase(@ptrCast(self), @ptrCast(it)));
    }

    pub fn Find(self: ?*anyopaque, key: []const u8) It::value_type {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return @ptrCast(qtc.QJsonObject_Find(@ptrCast(self), key_str));
    }

    pub fn Find2(self: ?*anyopaque, key: []const u8) QtC.QRect {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_Find2(@ptrCast(self), key_str);
    }

    pub fn ConstFind(self: ?*anyopaque, key: []const u8) QtC.QRect {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QJsonObject_ConstFind(@ptrCast(self), key_str);
    }

    pub fn Insert(self: ?*anyopaque, key: []const u8, value: ?*anyopaque) It::value_type {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return @ptrCast(qtc.QJsonObject_Insert(@ptrCast(self), key_str, @ptrCast(value)));
    }

    pub fn Empty(self: ?*anyopaque, ) bool {
        return qtc.QJsonObject_Empty(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonObject_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qjsonobject-iterator.html
pub const qjsonobject__iterator = struct {

    /// New constructs a new QJsonObject::iterator object.
    pub fn New() QtC.QJsonObject__iterator {
        return qtc.QJsonObject__iterator_new();
    }


    /// New2 constructs a new QJsonObject::iterator object.
    pub fn New2(obj: ?*anyopaque, index: i64) QtC.QJsonObject__iterator {
        return qtc.QJsonObject__iterator_new2(@ptrCast(obj), index);
    }


    pub fn Key(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QJsonObject__iterator_Key(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qjsonobject::iterator.Key: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn OperatorPlusPlus(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorPlusPlus(@ptrCast(self)));
    }

    pub fn OperatorPlusPlus2(self: ?*anyopaque, param1: i32) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorPlusPlus2(@ptrCast(self), param1));
    }

    pub fn OperatorMinusMinus(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorMinusMinus(@ptrCast(self)));
    }

    pub fn OperatorMinusMinus2(self: ?*anyopaque, param1: i32) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorMinusMinus2(@ptrCast(self), param1));
    }

    pub fn OperatorPlus(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorPlus(@ptrCast(self), j));
    }

    pub fn OperatorMinus(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorMinus(@ptrCast(self), j));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorPlusAssign(@ptrCast(self), j));
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonObject__iterator_OperatorMinusAssign(@ptrCast(self), j));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonObject__iterator_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qjsonobject-const_iterator.html
pub const qjsonobject__const_iterator = struct {

    /// New constructs a new QJsonObject::const_iterator object.
    pub fn New() QtC.QJsonObject__const_iterator {
        return qtc.QJsonObject__const_iterator_new();
    }


    /// New2 constructs a new QJsonObject::const_iterator object.
    pub fn New2(obj: ?*anyopaque, index: i64) QtC.QJsonObject__const_iterator {
        return qtc.QJsonObject__const_iterator_new2(@ptrCast(obj), index);
    }


    /// New3 constructs a new QJsonObject::const_iterator object.
    pub fn New3(other: *const It::value_type) QtC.QJsonObject__const_iterator {
        return qtc.QJsonObject__const_iterator_new3(@ptrCast(other));
    }


    pub fn Key(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QJsonObject__const_iterator_Key(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qjsonobject::const_iterator.Key: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn OperatorPlusPlus(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorPlusPlus(@ptrCast(self));
    }

    pub fn OperatorPlusPlus2(self: ?*anyopaque, param1: i32) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorPlusPlus2(@ptrCast(self), param1);
    }

    pub fn OperatorMinusMinus(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorMinusMinus(@ptrCast(self));
    }

    pub fn OperatorMinusMinus2(self: ?*anyopaque, param1: i32) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorMinusMinus2(@ptrCast(self), param1);
    }

    pub fn OperatorPlus(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorPlus(@ptrCast(self), j);
    }

    pub fn OperatorMinus(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorMinus(@ptrCast(self), j);
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorPlusAssign(@ptrCast(self), j);
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonObject__const_iterator_OperatorMinusAssign(@ptrCast(self), j);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonObject__const_iterator_Delete(@ptrCast(self));
    }
};

File: binding/src/libqjsonobject.cpp
#include <QDebug>
#include <QJsonObject>
#define WORKAROUND_INNER_CLASS_DEFINITION_QJsonObject__const_iterator
#define WORKAROUND_INNER_CLASS_DEFINITION_QJsonObject__iterator
#include <QJsonValue>
#include <QRect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStringView>
#include <qjsonobject.h>
#include "libqjsonobject.h"
#include "libqjsonobject.hxx"

QJsonObject* QJsonObject_new() {
	 return new QJsonObject();
}

QJsonObject* QJsonObject_new2(const QJsonObject* other) {
	 return new QJsonObject(*other);
}

void QJsonObject_OperatorAssign(QJsonObject* self, const QJsonObject* other) {
	self->operator=(*other);
}

void QJsonObject_Swap(QJsonObject* self, QJsonObject* other) {
	self->swap(*other);
}

QJsonObject* QJsonObject_FromVariantMap(const libqt_map mapVal) {
	return new QJsonObject(QJsonObject::fromVariantMap(*mapVal));
}

libqt_map QJsonObject_ToVariantMap(const QJsonObject* self) {
	return self->toVariantMap();
}

QJsonObject* QJsonObject_FromVariantHash(const libqt_map mapVal) {
	return new QJsonObject(QJsonObject::fromVariantHash(*mapVal));
}

libqt_map QJsonObject_ToVariantHash(const QJsonObject* self) {
	return self->toVariantHash();
}

libqt_list QJsonObject_Keys(const QJsonObject* self) {
	return self->keys();
}

ptrdiff_t QJsonObject_Size(const QJsonObject* self) {
	return self->size();
}

ptrdiff_t QJsonObject_Count(const QJsonObject* self) {
	return self->count();
}

ptrdiff_t QJsonObject_Length(const QJsonObject* self) {
	return self->length();
}

bool QJsonObject_IsEmpty(const QJsonObject* self) {
	return self->isEmpty();
}

QJsonValue* QJsonObject_Value(const QJsonObject* self, const libqt_string key) {
	return new QJsonValue(self->value(QString::fromUtf8(key.data, key.len)));
}

QJsonValue* QJsonObject_OperatorSubscript(const QJsonObject* self, const libqt_string key) {
	return new QJsonValue(self->operator[](QString::fromUtf8(key.data, key.len)));
}

QJsonValue* QJsonObject_Value2(const QJsonObject* self, QStringView* key) {
	return new QJsonValue(self->value(*key));
}

QJsonValue* QJsonObject_Value3(const QJsonObject* self, QLatin1StringView key) {
	return new QJsonValue(self->value(key));
}

QJsonValue* QJsonObject_OperatorSubscript3(const QJsonObject* self, QStringView* key) {
	return new QJsonValue(self->operator[](*key));
}

QJsonValue* QJsonObject_OperatorSubscript4(const QJsonObject* self, QLatin1StringView key) {
	return new QJsonValue(self->operator[](key));
}

void QJsonObject_Remove(QJsonObject* self, const libqt_string key) {
	self->remove(QString::fromUtf8(key.data, key.len));
}

QJsonValue* QJsonObject_Take(QJsonObject* self, const libqt_string key) {
	return new QJsonValue(self->take(QString::fromUtf8(key.data, key.len)));
}

bool QJsonObject_Contains(const QJsonObject* self, const libqt_string key) {
	return self->contains(QString::fromUtf8(key.data, key.len));
}

void QJsonObject_Remove2(QJsonObject* self, QStringView* key) {
	self->remove(*key);
}

void QJsonObject_Remove3(QJsonObject* self, QLatin1StringView key) {
	self->remove(key);
}

QJsonValue* QJsonObject_Take2(QJsonObject* self, QStringView* key) {
	return new QJsonValue(self->take(*key));
}

QJsonValue* QJsonObject_Take3(QJsonObject* self, QLatin1StringView key) {
	return new QJsonValue(self->take(key));
}

bool QJsonObject_Contains2(const QJsonObject* self, QStringView* key) {
	return self->contains(*key);
}

bool QJsonObject_Contains3(const QJsonObject* self, QLatin1StringView key) {
	return self->contains(key);
}

It::value_type* QJsonObject_Begin(QJsonObject* self) {
	return self->begin();
}

const QRect* QJsonObject_Begin2(const QJsonObject* self) {
	return self->begin();
}

const QRect* QJsonObject_ConstBegin(const QJsonObject* self) {
	return self->constBegin();
}

It::value_type* QJsonObject_End(QJsonObject* self) {
	return self->end();
}

const QRect* QJsonObject_End2(const QJsonObject* self) {
	return self->end();
}

const QRect* QJsonObject_ConstEnd(const QJsonObject* self) {
	return self->constEnd();
}

It::value_type* QJsonObject_Erase(QJsonObject* self, It::value_type* it) {
	return self->erase(it);
}

It::value_type* QJsonObject_Find(QJsonObject* self, const libqt_string key) {
	return self->find(QString::fromUtf8(key.data, key.len));
}

const QRect* QJsonObject_Find2(const QJsonObject* self, const libqt_string key) {
	return self->find(QString::fromUtf8(key.data, key.len));
}

const QRect* QJsonObject_ConstFind(const QJsonObject* self, const libqt_string key) {
	return self->constFind(QString::fromUtf8(key.data, key.len));
}

It::value_type* QJsonObject_Insert(QJsonObject* self, const libqt_string key, const QJsonValue* value) {
	return self->insert(QString::fromUtf8(key.data, key.len), *value);
}

It::value_type* QJsonObject_Find3(QJsonObject* self, QStringView* key) {
	return self->find(*key);
}

It::value_type* QJsonObject_Find4(QJsonObject* self, QLatin1StringView key) {
	return self->find(key);
}

const QRect* QJsonObject_Find5(const QJsonObject* self, QStringView* key) {
	return self->find(*key);
}

const QRect* QJsonObject_Find6(const QJsonObject* self, QLatin1StringView key) {
	return self->find(key);
}

const QRect* QJsonObject_ConstFind2(const QJsonObject* self, QStringView* key) {
	return self->constFind(*key);
}

const QRect* QJsonObject_ConstFind3(const QJsonObject* self, QLatin1StringView key) {
	return self->constFind(key);
}

It::value_type* QJsonObject_Insert2(QJsonObject* self, QStringView* key, const QJsonValue* value) {
	return self->insert(*key, *value);
}

It::value_type* QJsonObject_Insert3(QJsonObject* self, QLatin1StringView key, const QJsonValue* value) {
	return self->insert(key, *value);
}

bool QJsonObject_Empty(const QJsonObject* self) {
	return self->empty();
}

void QJsonObject_Delete(QJsonObject* self) {
    delete self;
}

QJsonObject__iterator* QJsonObject__iterator_new(const QJsonObject__iterator* other) {
	 return new QJsonObject::iterator(*other);
}

QJsonObject__iterator* QJsonObject__iterator_new2() {
	 return new QJsonObject::iterator();
}

QJsonObject__iterator* QJsonObject__iterator_new3(QJsonObject* obj, ptrdiff_t index) {
	 return new QJsonObject::iterator(obj, index);
}

QJsonObject__iterator* QJsonObject__iterator_new4(const QJsonObject__iterator* other) {
	 return new QJsonObject::iterator(*other);
}

void QJsonObject__iterator_OperatorAssign(QJsonObject__iterator* self, const QJsonObject__iterator* other) {
	self->operator=(*other);
}

libqt_string QJsonObject__iterator_Key(const QJsonObject__iterator* self) {
	QString _ret = self->key();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

It::value_type* QJsonObject__iterator_OperatorPlusPlus(QJsonObject__iterator* self) {
	return self->operator++();
}

It::value_type* QJsonObject__iterator_OperatorPlusPlus2(QJsonObject__iterator* self, int param1) {
	return self->operator++(param1);
}

It::value_type* QJsonObject__iterator_OperatorMinusMinus(QJsonObject__iterator* self) {
	return self->operator--();
}

It::value_type* QJsonObject__iterator_OperatorMinusMinus2(QJsonObject__iterator* self, int param1) {
	return self->operator--(param1);
}

It::value_type* QJsonObject__iterator_OperatorPlus(const QJsonObject__iterator* self, ptrdiff_t j) {
	return self->operator+(j);
}

It::value_type* QJsonObject__iterator_OperatorMinus(const QJsonObject__iterator* self, ptrdiff_t j) {
	return self->operator-(j);
}

It::value_type* QJsonObject__iterator_OperatorPlusAssign(QJsonObject__iterator* self, ptrdiff_t j) {
	return self->operator+=(j);
}

It::value_type* QJsonObject__iterator_OperatorMinusAssign(QJsonObject__iterator* self, ptrdiff_t j) {
	return self->operator-=(j);
}

ptrdiff_t QJsonObject__iterator_OperatorMinus2(const QJsonObject__iterator* self, QJsonObject__iterator* j) {
	return self->operator-(*j);
}

void QJsonObject__iterator_Delete(QJsonObject__iterator* self) {
    delete self;
}

QJsonObject__const_iterator* QJsonObject__const_iterator_new(const QJsonObject__const_iterator* other) {
	 return new QJsonObject::const_iterator(*other);
}

QJsonObject__const_iterator* QJsonObject__const_iterator_new2() {
	 return new QJsonObject::const_iterator();
}

QJsonObject__const_iterator* QJsonObject__const_iterator_new3(const QJsonObject* obj, ptrdiff_t index) {
	 return new QJsonObject::const_iterator(obj, index);
}

QJsonObject__const_iterator* QJsonObject__const_iterator_new4(const It::value_type* other) {
	 return new QJsonObject::const_iterator(other);
}

QJsonObject__const_iterator* QJsonObject__const_iterator_new5(const QJsonObject__const_iterator* other) {
	 return new QJsonObject::const_iterator(*other);
}

void QJsonObject__const_iterator_OperatorAssign(QJsonObject__const_iterator* self, const QJsonObject__const_iterator* other) {
	self->operator=(*other);
}

libqt_string QJsonObject__const_iterator_Key(const QJsonObject__const_iterator* self) {
	QString _ret = self->key();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

const QRect* QJsonObject__const_iterator_OperatorPlusPlus(QJsonObject__const_iterator* self) {
	const_iterator& _ret = self->operator++();
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonObject__const_iterator_OperatorPlusPlus2(QJsonObject__const_iterator* self, int param1) {
	return self->operator++(param1);
}

const QRect* QJsonObject__const_iterator_OperatorMinusMinus(QJsonObject__const_iterator* self) {
	const_iterator& _ret = self->operator--();
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonObject__const_iterator_OperatorMinusMinus2(QJsonObject__const_iterator* self, int param1) {
	return self->operator--(param1);
}

const QRect* QJsonObject__const_iterator_OperatorPlus(const QJsonObject__const_iterator* self, ptrdiff_t j) {
	return self->operator+(j);
}

const QRect* QJsonObject__const_iterator_OperatorMinus(const QJsonObject__const_iterator* self, ptrdiff_t j) {
	return self->operator-(j);
}

const QRect* QJsonObject__const_iterator_OperatorPlusAssign(QJsonObject__const_iterator* self, ptrdiff_t j) {
	const_iterator& _ret = self->operator+=(j);
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonObject__const_iterator_OperatorMinusAssign(QJsonObject__const_iterator* self, ptrdiff_t j) {
	const_iterator& _ret = self->operator-=(j);
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

ptrdiff_t QJsonObject__const_iterator_OperatorMinus2(const QJsonObject__const_iterator* self, QJsonObject__const_iterator* j) {
	return self->operator-(*j);
}

void QJsonObject__const_iterator_Delete(QJsonObject__const_iterator* self) {
    delete self;
}

